File: solution1.py
# https://leetcode.com/problems/reverse-linked-list/
# // Time Complexity : o(n) recursively visits all node
# // Space Complexity : o(length of input) due to stack
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this :no
#
#
# // Your code here along with comments explaining your approach
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return head
        return self.helper(head, None)

    def helper(self, curr, prev):
        if not curr:
            return prev
        head = self.helper(curr.next, curr)
        curr.next = prev
        return head

File: Solution1.java
//Remove Nth Node From End of List

//TC:O(n)
//SC:O(1)
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution1 {
    public ListNode removeNthFromEnd(ListNode head, int n) {

        ListNode dummy=new ListNode(-1);
        dummy.next=head;
        ListNode fast=dummy;
        int count=0;
      
        while(count<=n)
        {
            fast=fast.next;
            count++;
        }
       ListNode slow=dummy;
        while(fast!=null)
        {
            slow=slow.next;
            fast=fast.next;
        }
        //remove node next to slow
        ListNode temp=slow.next;
        slow.next=slow.next.next;
        temp.next=null;
        return dummy.next;
     
    }
}